File: internal/gateway/dns.go
package gateway

import (
	"context"
	"fmt"
	"net"
	"net/netip"
	"strings"
	"sync"
	"time"

	"github.com/miekg/dns"
)

type DNSServer struct {
	config  DNSConfig
	servers []*dns.Server
	mu      sync.Mutex
	running bool
	cidrs   []netip.Prefix
}

type DNSDebugResult struct {
	Name     string   `json:"name"`
	Type     string   `json:"type"`
	Upstream string   `json:"upstream"`
	Answers  []string `json:"answers"`
	Detail   string   `json:"detail,omitempty"`
}

func NewDNSServer(config DNSConfig) (*DNSServer, error) {
	config.Enabled = true
	cidrs := []netip.Prefix{}
	for _, value := range config.DomesticCIDRs {
		prefix, err := netip.ParsePrefix(value)
		if err != nil {
			return nil, err
		}
		cidrs = append(cidrs, prefix)
	}
	return &DNSServer{config: config, cidrs: cidrs}, nil
}

func (server *DNSServer) Start() error {
	server.mu.Lock()
	defer server.mu.Unlock()
	if server.running {
		return nil
	}
	handler := dns.HandlerFunc(server.handle)
	for _, host := range server.config.ListenHosts {
		address := net.JoinHostPort(host, fmt.Sprint(server.config.ListenPort))
		for _, network := range []string{"udp", "tcp"} {
			current := &dns.Server{Addr: address, Net: network, Handler: handler, ReadTimeout: 5 * time.Second, WriteTimeout: 5 * time.Second}
			if network == "udp" {
				packet, err := net.ListenPacket("udp4", address)
				if err != nil {
					_ = server.stopUnlocked()
					return fmt.Errorf("listen DNS %s: %w", address, err)
				}
				current.PacketConn = packet
			} else {
				listener, err := net.Listen("tcp4", address)
				if err != nil {
					_ = server.stopUnlocked()
					return fmt.Errorf("listen DNS %s: %w", address, err)
				}
				current.Listener = listener
			}
			server.servers = append(server.servers, current)
			go func(item *dns.Server) { _ = item.ActivateAndServe() }(current)
		}
	}
	server.running = true
	return nil
}

func (server *DNSServer) Stop() error {
	server.mu.Lock()
	defer server.mu.Unlock()
	return server.stopUnlocked()
}

func (server *DNSServer) stopUnlocked() error {
	var failure error
	for _, current := range server.servers {
		if err := current.Shutdown(); err != nil && failure == nil {
			failure = err
		}
	}
	server.servers = nil
	server.running = false
	return failure
}

func (server *DNSServer) Running() bool {
	server.mu.Lock()
	defer server.mu.Unlock()
	return server.running
}

func (server *DNSServer) handle(writer dns.ResponseWriter, request *dns.Msg) {
	response, upstream, err := server.resolve(context.Background(), request)
	if err != nil {
		response = new(dns.Msg)
		response.SetRcode(request, dns.RcodeServerFailure)
		response.RecursionAvailable = true
	} else {
		_ = upstream
	}
	_ = writer.WriteMsg(response)
}

func (server *DNSServer) DebugQuery(ctx context.Context, name, recordType string) (DNSDebugResult, error) {
	queryType := dns.StringToType[strings.ToUpper(valueOr(recordType, "A"))]
	if queryType == 0 {
		return DNSDebugResult{}, fmt.Errorf("unsupported DNS query type %q", recordType)
	}
	message := new(dns.Msg)
	message.SetQuestion(dns.Fqdn(name), queryType)
	response, upstream, err := server.resolve(ctx, message)
	if err != nil {
		return DNSDebugResult{}, err
	}
	answers := []string{}
	for _, answer := range response.Answer {
		answers = append(answers, answer.String())
	}
	return DNSDebugResult{Name: dns.Fqdn(name), Type: dns.TypeToString[queryType], Upstream: upstream, Answers: answers}, nil
}

func (server *DNSServer) resolve(ctx context.Context, request *dns.Msg) (*dns.Msg, string, error) {
	if len(request.Question) == 0 {
		response := new(dns.Msg)
		response.SetReply(request)
		return response, "", nil
	}
	question := request.Question[0]
	if server.config.RejectHTTPS && question.Qtype == dns.TypeHTTPS {
		response := new(dns.Msg)
		response.SetRcode(request, dns.RcodeNameError)
		response.RecursionAvailable = true
		return response, "reject", nil
	}
	if upstream := server.matchRuleUpstream(question.Name); upstream != "" {
		return server.exchange(ctx, request, server.namedUpstream(upstream))
	}
	if server.config.Strategy == DNSStrategyRulesFirst {
		return server.exchange(ctx, request, server.config.RemoteUpstream)
	}
	localResponse, upstream, localErr := server.exchange(ctx, request, server.firstLocalUpstream())
	if localErr == nil && server.domesticResponse(localResponse) {
		return localResponse, upstream, nil
	}
	remoteResponse, remote, remoteErr := server.exchange(ctx, request, server.config.RemoteUpstream)
	if remoteErr == nil {
		return remoteResponse, remote, nil
	}
	if localErr == nil {
		return localResponse, upstream, nil
	}
	return nil, "", remoteErr
}

func (server *DNSServer) exchange(ctx context.Context, request *dns.Msg, upstream string) (*dns.Msg, string, error) {
	client := &dns.Client{Net: "udp", Timeout: 5 * time.Second}
	done := make(chan struct {
		msg *dns.Msg
		err error
	}, 1)
	go func() {
		msg, _, err := client.Exchange(request.Copy(), upstream)
		done <- struct {
			msg *dns.Msg
			err error
		}{msg: msg, err: err}
	}()
	select {
	case <-ctx.Done():
		return nil, upstream, ctx.Err()
	case result := <-done:
		return result.msg, upstream, result.err
	}
}

func (server *DNSServer) firstLocalUpstream() string {
	if len(server.config.LocalUpstreams) == 0 {
		return server.config.RemoteUpstream
	}
	return server.config.LocalUpstreams[0]
}

func (server *DNSServer) namedUpstream(value string) string {
	switch value {
	case "local":
		return server.firstLocalUpstream()
	case "remote", "":
		return server.config.RemoteUpstream
	default:
		return value
	}
}

func (server *DNSServer) matchRuleUpstream(name string) string {
	name = strings.TrimSuffix(strings.ToLower(name), ".")
	for _, set := range server.config.RuleSets {
		if !set.Enabled {
			continue
		}
		for _, rule := range set.Rules {
			if ruleMatchesDomain(rule, name) {
				return set.Upstream
			}
		}
	}
	return ""
}

func ruleMatchesDomain(rule, name string) bool {
	rule = strings.TrimSpace(strings.ToLower(rule))
	if rule == "" || strings.HasPrefix(rule, "#") {
		return false
	}
	kind, payload, ok := strings.Cut(rule, ",")
	if !ok {
		payload = rule
		kind = "domain-suffix"
	}
	payload = strings.TrimSuffix(strings.TrimSpace(payload), ".")
	switch strings.TrimSpace(kind) {
	case "domain":
		return name == payload
	case "domain-keyword":
		return strings.Contains(name, payload)
	default:
		return name == payload || strings.HasSuffix(name, "."+payload)
	}
}

func (server *DNSServer) domesticResponse(message *dns.Msg) bool {
	if message == nil || len(message.Answer) == 0 {
		return false
	}
	for _, answer := range message.Answer {
		var addr netip.Addr
		switch value := answer.(type) {
		case *dns.A:
			addr, _ = netip.AddrFromSlice(value.A)
		default:
			continue
		}
		for _, prefix := range server.cidrs {
			if prefix.Contains(addr) {
				return true
			}
		}
	}
	return false
}

File: rust/crates/sempre-gateway/src/dns.rs
use std::{net::Ipv4Addr, sync::Arc, time::Duration};

use serde::Serialize;
use tokio::{
    io::{AsyncReadExt as _, AsyncWriteExt as _},
    net::{TcpListener, TcpStream, UdpSocket, lookup_host},
    sync::watch,
    task::{JoinHandle, JoinSet},
    time::timeout,
};

use crate::model::DnsConfig;
use crate::{
    GatewayError,
    dns_wire::{
        TYPE_HTTPS, answer_ipv4_addresses, build_query, format_answers, fqdn, parse_question,
        record_number, response_with_code,
    },
    domain_matcher::DomainMatcher,
};

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct DnsDebugResult {
    pub name: String,
    #[serde(rename = "type")]
    pub record_type: String,
    pub upstream: String,
    pub answers: Vec<String>,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub detail: String,
}

pub(crate) struct DnsServer {
    shutdown: watch::Sender<bool>,
    tasks: Vec<JoinHandle<()>>,
}

#[derive(Clone)]
struct Resolver {
    config: DnsConfig,
    domestic_cidrs: Vec<(Ipv4Addr, u8)>,
    rule_sets: Vec<ResolvedRuleSet>,
}

#[derive(Clone)]
struct ResolvedRuleSet {
    id: String,
    upstream: String,
    matcher: DomainMatcher,
}

struct Resolved {
    packet: Vec<u8>,
    upstream: String,
    detail: String,
}

impl DnsServer {
    pub(crate) async fn start(config: DnsConfig) -> Result<Self, GatewayError> {
        let resolver = Resolver::new(config)?;
        let mut udp = Vec::new();
        let mut tcp = Vec::new();
        for host in &resolver.config.listen_hosts {
            let address = format!("{host}:{}", resolver.config.listen_port);
            udp.push(
                UdpSocket::bind(&address).await.map_err(|error| {
                    GatewayError::io(format!("listen DNS UDP {address}"), error)
                })?,
            );
            tcp.push(
                TcpListener::bind(&address).await.map_err(|error| {
                    GatewayError::io(format!("listen DNS TCP {address}"), error)
                })?,
            );
        }
        let (shutdown, _) = watch::channel(false);
        let mut tasks = Vec::with_capacity(udp.len() + tcp.len());
        for socket in udp {
            tasks.push(tokio::spawn(serve_udp(
                socket,
                resolver.clone(),
                shutdown.subscribe(),
            )));
        }
        for listener in tcp {
            tasks.push(tokio::spawn(serve_tcp(
                listener,
                resolver.clone(),
                shutdown.subscribe(),
            )));
        }
        Ok(Self { shutdown, tasks })
    }

    pub(crate) async fn stop(self) {
        let _ = self.shutdown.send(true);
        for task in self.tasks {
            let _ = task.await;
        }
    }
}

pub(crate) async fn debug_query(
    config: DnsConfig,
    name: &str,
    record_type: &str,
) -> Result<DnsDebugResult, GatewayError> {
    Resolver::new(config)?.debug_query(name, record_type).await
}

pub fn managed_probe_names(config: &DnsConfig) -> Result<(String, String), GatewayError> {
    let resolver = Resolver::new(config.clone())?;
    let local = [
        "baidu.com",
        "qq.com",
        "taobao.com",
        "jd.com",
        "bilibili.com",
    ]
    .into_iter()
    .find(|name| resolver.selected_upstream(name) == "local")
    .ok_or_else(|| GatewayError::invalid("managed DNS policy has no local probe domain"))?;
    let remote = [
        "example.com",
        "github.com",
        "wikipedia.org",
        "google.com",
        "youtube.com",
    ]
    .into_iter()
    .find(|name| resolver.selected_upstream(name) == "remote")
    .ok_or_else(|| GatewayError::invalid("managed DNS policy has no core probe domain"))?;
    Ok((local.into(), remote.into()))
}

impl Resolver {
    fn new(config: DnsConfig) -> Result<Self, GatewayError> {
        let domestic_cidrs = config
            .domestic_cidrs
            .iter()
            .map(|value| parse_cidr(value))
            .collect::<Result<Vec<_>, _>>()?;
        let rule_sets = config
            .rule_sets
            .iter()
            .filter(|rule_set| rule_set.enabled)
            .map(|rule_set| ResolvedRuleSet {
                id: rule_set.id.clone(),
                upstream: rule_set.upstream.clone(),
                matcher: DomainMatcher::from_rules(&rule_set.rules),
            })
            .collect();
        Ok(Self {
            config,
            domestic_cidrs,
            rule_sets,
        })
    }

    async fn debug_query(
        &self,
        name: &str,
        record_type: &str,
    ) -> Result<DnsDebugResult, GatewayError> {
        let (record_type, number) = record_number(record_type)?;
        let packet = build_query(name, number)?;
        let resolved = self.resolve(&packet).await?;
        Ok(DnsDebugResult {
            name: fqdn(name),
            record_type,
            upstream: resolved.upstream,
            answers: format_answers(&resolved.packet)?,
            detail: resolved.detail,
        })
    }

    async fn resolve(&self, packet: &[u8]) -> Result<Resolved, GatewayError> {
        let question = parse_question(packet)?;
        let Some(question) = question else {
            return Ok(Resolved {
                packet: response_with_code(packet, 0)?,
                upstream: String::new(),
                detail: "empty-question".into(),
            });
        };
        if self.config.reject_https && question.record_type == TYPE_HTTPS {
            return Ok(Resolved {
                packet: response_with_code(packet, 3)?,
                upstream: "reject".into(),
                detail: "https-rejected".into(),
            });
        }
        if let Some(rule_set) = self.match_rule(&question.name) {
            let mut resolved = self.exchange_named(packet, &rule_set.upstream).await?;
            resolved.detail = format!("rule-set:{}", rule_set.id);
            return Ok(resolved);
        }
        if self.config.strategy == "rules-first" {
            let mut resolved = self.exchange(packet, &self.config.remote_upstream).await?;
            resolved.detail = "default-remote".into();
            return Ok(resolved);
        }
        let local = self.exchange_local(packet).await;
        if local
            .as_ref()
            .is_ok_and(|response| self.domestic_response(&response.packet))
        {
            return local.map(|mut response| {
                response.detail = "local-response".into();
                response
            });
        }
        match self.exchange(packet, &self.config.remote_upstream).await {
            Ok(mut remote) => {
                remote.detail = "remote-response".into();
                Ok(remote)
            }
            Err(remote_error) => local.or(Err(remote_error)),
        }
    }

    async fn exchange_named(
        &self,
        packet: &[u8],
        upstream: &str,
    ) -> Result<Resolved, GatewayError> {
        match upstream {
            "local" => self.exchange_local(packet).await,
            "remote" | "" => self.exchange(packet, &self.config.remote_upstream).await,
            upstream => self.exchange(packet, upstream).await,
        }
    }

    async fn exchange_local(&self, packet: &[u8]) -> Result<Resolved, GatewayError> {
        if self.config.local_upstreams.is_empty() {
            return self.exchange(packet, &self.config.remote_upstream).await;
        }
        let mut last_error = None;
        for upstream in &self.config.local_upstreams {
            match self.exchange(packet, upstream).await {
                Ok(response) => return Ok(response),
                Err(error) => last_error = Some(error),
            }
        }
        Err(last_error.expect("non-empty local upstreams"))
    }

    async fn exchange(&self, packet: &[u8], upstream: &str) -> Result<Resolved, GatewayError> {
        let address = lookup_host(upstream)
            .await
            .map_err(|error| GatewayError::io(format!("resolve DNS upstream {upstream}"), error))?
            .next()
            .ok_or_else(|| GatewayError::invalid(format!("DNS upstream {upstream:?} is empty")))?;
        let socket = UdpSocket::bind("0.0.0.0:0")
            .await
            .map_err(|error| GatewayError::io("bind DNS upstream socket", error))?;
        socket
            .connect(address)
            .await
            .map_err(|error| GatewayError::io(format!("connect DNS upstream {upstream}"), error))?;
        socket
            .send(packet)
            .await
            .map_err(|error| GatewayError::io(format!("send DNS query to {upstream}"), error))?;
        let mut response = vec![0_u8; u16::MAX as usize];
        let count = timeout(Duration::from_secs(5), socket.recv(&mut response))
            .await
            .map_err(|_| GatewayError::invalid(format!("DNS upstream {upstream} timed out")))?
            .map_err(|error| {
                GatewayError::io(format!("receive DNS response from {upstream}"), error)
            })?;
        response.truncate(count);
        if response.get(..2) != packet.get(..2) {
            return Err(GatewayError::invalid(format!(
                "DNS upstream {upstream} returned a mismatched transaction"
            )));
        }
        Ok(Resolved {
            packet: response,
            upstream: upstream.into(),
            detail: String::new(),
        })
    }

    fn match_rule(&self, name: &str) -> Option<&ResolvedRuleSet> {
        self.rule_sets
            .iter()
            .find(|rule_set| rule_set.matcher.matches(name))
    }

    fn selected_upstream(&self, name: &str) -> &str {
        self.match_rule(name)
            .map_or("remote", |rule_set| rule_set.upstream.as_str())
    }

    fn domestic_response(&self, packet: &[u8]) -> bool {
        answer_ipv4_addresses(packet).is_ok_and(|addresses| {
            addresses.into_iter().any(|address| {
                self.domestic_cidrs
                    .iter()
                    .any(|(network, prefix)| in_prefix(address, *network, *prefix))
            })
        })
    }
}

async fn serve_udp(socket: UdpSocket, resolver: Resolver, mut shutdown: watch::Receiver<bool>) {
    let socket = Arc::new(socket);
    let mut buffer = vec![0_u8; u16::MAX as usize];
    let mut requests = JoinSet::new();
    loop {
        tokio::select! {
            changed = shutdown.changed() => {
                if changed.is_err() || *shutdown.borrow() {
                    requests.abort_all();
                    while requests.join_next().await.is_some() {}
                    return;
                }
            }
            Some(_) = requests.join_next(), if !requests.is_empty() => {}
            result = socket.recv_from(&mut buffer) => {
                let Ok((count, peer)) = result else { return; };
                let request = buffer[..count].to_vec();
                let socket = Arc::clone(&socket);
                let resolver = resolver.clone();
                requests.spawn(async move {
                    let response = resolver.resolve(&request).await
                        .map_or_else(|_| response_with_code(&request, 2).unwrap_or_default(), |value| value.packet);
                    if !response.is_empty() { let _ = socket.send_to(&response, peer).await; }
                });
            }
        }
    }
}

async fn serve_tcp(listener: TcpListener, resolver: Resolver, mut shutdown: watch::Receiver<bool>) {
    let mut connections = JoinSet::new();
    loop {
        tokio::select! {
            changed = shutdown.changed() => {
                if changed.is_err() || *shutdown.borrow() {
                    connections.abort_all();
                    while connections.join_next().await.is_some() {}
                    return;
                }
            }
            Some(_) = connections.join_next(), if !connections.is_empty() => {}
            accepted = listener.accept() => {
                let Ok((stream, _)) = accepted else { return; };
                let resolver = resolver.clone();
                connections.spawn(async move { let _ = serve_tcp_connection(stream, resolver).await; });
            }
        }
    }
}

async fn serve_tcp_connection(
    mut stream: TcpStream,
    resolver: Resolver,
) -> Result<(), GatewayError> {
    let mut length = [0_u8; 2];
    timeout(Duration::from_secs(5), stream.read_exact(&mut length))
        .await
        .map_err(|_| GatewayError::invalid("DNS TCP read timed out"))?
        .map_err(|error| GatewayError::io("read DNS TCP length", error))?;
    let mut request = vec![0_u8; usize::from(u16::from_be_bytes(length))];
    stream
        .read_exact(&mut request)
        .await
        .map_err(|error| GatewayError::io("read DNS TCP query", error))?;
    let response = resolver.resolve(&request).await.map_or_else(
        |_| response_with_code(&request, 2).unwrap_or_default(),
        |value| value.packet,
    );
    let length = u16::try_from(response.len())
        .map_err(|_| GatewayError::invalid("DNS TCP response exceeds 65535 bytes"))?;
    stream
        .write_all(&length.to_be_bytes())
        .await
        .map_err(|error| GatewayError::io("write DNS TCP response length", error))?;
    stream
        .write_all(&response)
        .await
        .map_err(|error| GatewayError::io("write DNS TCP response", error))
}

fn parse_cidr(value: &str) -> Result<(Ipv4Addr, u8), GatewayError> {
    let (address, prefix) = value
        .split_once('/')
        .ok_or_else(|| GatewayError::invalid(format!("invalid IPv4 prefix {value:?}")))?;
    let address = address
        .parse()
        .map_err(|_| GatewayError::invalid(format!("invalid IPv4 prefix {value:?}")))?;
    let prefix = prefix
        .parse::<u8>()
        .ok()
        .filter(|prefix| *prefix <= 32)
        .ok_or_else(|| GatewayError::invalid(format!("invalid IPv4 prefix {value:?}")))?;
    Ok((address, prefix))
}

fn in_prefix(address: Ipv4Addr, network: Ipv4Addr, prefix: u8) -> bool {
    let mask = if prefix == 0 {
        0
    } else {
        u32::MAX << (32 - prefix)
    };
    u32::from(address) & mask == u32::from(network) & mask
}

#[cfg(test)]
mod tests {
    use super::*;

    async fn answering_upstream(count: usize, address: [u8; 4]) -> (String, JoinHandle<()>) {
        let upstream = UdpSocket::bind("127.0.0.1:0").await.expect("upstream");
        let socket_address = upstream.local_addr().expect("upstream address");
        let responder = tokio::spawn(async move {
            for _ in 0..count {
                let mut query = [0_u8; 512];
                let (count, peer) = upstream.recv_from(&mut query).await.expect("query");
                let mut response = query[..count].to_vec();
                response[2] |= 0x80;
                response[3] |= 0x80;
                response[6..8].copy_from_slice(&1_u16.to_be_bytes());
                response.extend_from_slice(&[
                    0xc0, 0x0c, 0, 1, 0, 1, 0, 0, 0, 60, 0, 4, address[0], address[1], address[2],
                    address[3],
                ]);
                upstream.send_to(&response, peer).await.expect("response");
            }
        });
        (socket_address.to_string(), responder)
    }

    #[tokio::test]
    async fn debug_query_exchanges_with_the_selected_udp_upstream() {
        let upstream = UdpSocket::bind("127.0.0.1:0").await.expect("upstream");
        let address = upstream.local_addr().expect("upstream address");
        let responder = tokio::spawn(async move {
            let mut query = [0_u8; 512];
            let (count, peer) = upstream.recv_from(&mut query).await.expect("query");
            let mut response = query[..count].to_vec();
            response[2] |= 0x80;
            response[3] |= 0x80;
            response[6..8].copy_from_slice(&1_u16.to_be_bytes());
            response.extend_from_slice(&[0xc0, 0x0c, 0, 1, 0, 1, 0, 0, 0, 60, 0, 4, 10, 0, 0, 1]);
            upstream.send_to(&response, peer).await.expect("response");
        });
        let config = DnsConfig {
            local_upstreams: vec![address.to_string()],
            remote_upstream: address.to_string(),
            ..DnsConfig::default()
        };
        let result = debug_query(config, "example.com", "A")
            .await
            .expect("debug query");
        responder.await.expect("responder");
        assert_eq!(result.upstream, address.to_string());
        assert!(result.answers[0].ends_with("A 10.0.0.1"));
    }

    #[tokio::test]
    async fn managed_frontend_enforces_proxy_direct_domestic_then_default_order() {
        let (local, local_task) = answering_upstream(2, [10, 0, 0, 1]).await;
        let (remote, remote_task) = answering_upstream(2, [198, 18, 0, 1]).await;
        let config = DnsConfig::managed_frontend(
            1054,
            vec![local],
            remote,
            vec!["domain,proxy.baidu.com".into()],
            vec!["domain,direct.example".into()],
            false,
        )
        .expect("managed frontend");
        for (name, answer, detail) in [
            ("proxy.baidu.com", "198.18.0.1", "rule-set:explicit-proxy"),
            ("direct.example", "10.0.0.1", "rule-set:explicit-direct"),
            ("baidu.com", "10.0.0.1", "rule-set:domestic-domains"),
            ("github.com", "198.18.0.1", "default-remote"),
        ] {
            let result = debug_query(config.clone(), name, "A")
                .await
                .expect("debug query");
            assert!(result.answers[0].ends_with(&format!("A {answer}")));
            assert_eq!(result.detail, detail);
        }
        local_task.await.expect("local responder");
        remote_task.await.expect("remote responder");
    }
}
